feat: rich link component

surajsharma opened the pull request:
#371 
eta 4 hours

markholdex commented:
@surajsharma any updates here?

markholdex commented:
ETA was 4 hours. Did you encounter a blocker?

surajsharma commented:
> ETA was 4 hours. Did you encounter a blocker?

no blockers, just occupied with #493 and #496 , will wrap this up presently with the others. eta remains the same

surajsharma commented:
@markholdex @georgeciubotaru pls give access to vercel, i cannot inspect the failing build

georgeciubotaru commented:
@surajsharma 

You can run `npm run build` and then `npm run preview` and test yourself on your local machine
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/97b1a796-5631-4255-8b7c-e9f6f6805d4b)

surajsharma commented:
> @surajsharma
> 
> You can run `npm run build` and then `npm run preview` and test yourself on your local machine ![image](https://user-images.githubusercontent.com/8217518/265338312-97b1a796-5631-4255-8b7c-e9f6f6805d4b.png)

both `build` and `preview` are running fine on my end @georgeciubotaru

surajsharma commented:
@predrag-us @georgeciubotaru also, i think there should be some contingency here for cases where the og calls fails/returns incomplete data...the design has 3 things

<img width="620" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/4b6b3a04-1f56-4e4a-90d7-02657b5878b8">


from top to bottom we have:

1) Title 
2) Description
3) truncated link

the truncated link isn't from OG data, but i assume the other two are.

please mention if there needs to be some special text if the og api call fails

georgeciubotaru commented:
@surajsharma 

1. Update the repo to use `node:18`
2. Resolve conflicts
3. Ping me to update the node version in Vercel

surajsharma commented:
@georgeciubotaru commit resolved, node is `18.16` on local, pls update on vercel

georgeciubotaru commented on src/lib/components/BodyRenderer/item.pug:
```diff
@@ -38,7 +38,21 @@
+
+    Case(value="linkTool")
+      RichLink(item="{item}")
+      //- em(style:color="red") 
+      //-   | {item.data.url}
+      //-   .br
+      //-   | {item.data.embed}
+        //- .hidden
+          //- | {console.log(item)}
```
why don't you remove the logs?

surajsharma replied:
@georgeciubotaru sorry, removing now 😰

georgeciubotaru commented on src/routes/api/link/+server.ts:
```diff
@@ -0,0 +1,34 @@
+            image: ogImage,
+            name: ogSiteName,
+          },
+          {
```
set a caching value for 5 mins

surajsharma replied:
done

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/richLink.svelte:
```diff
@@ -0,0 +1,88 @@
+          target={isHoldexLink ? '_self' : '_blank'}
+          rel="noreferrer"
+        >
+          <slot>
```
you can obtain truncation via CSS

surajsharma replied:
@georgeciubotaru done

georgeciubotaru replied:
Also, why do you need `slot`?

surajsharma replied:
@georgeciubotaru slot removed

georgeciubotaru commented:
@surajsharma See the Vercel Preview comments

surajsharma commented:
@georgeciubotaru don't have access 
<img width="447" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/d9c23997-2894-409f-80aa-58838fa73f5a">

georgeciubotaru commented:
@surajsharma 

- [ ] - 
<img width="520" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/8217518/996ad527-7d72-4249-a3c1-2566065b010e">

- [ ] - 
<img width="510" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/8217518/716b5d4a-9810-49ff-8d0f-507ff0405a73">
- [ ] - 
<img width="571" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/8217518/c74194f9-73ae-495f-9a0b-adf7a020d4c3">
- [ ] - 
<img width="984" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/8217518/f51cd738-c97f-4ab2-96c3-9b7f8f37f4b7">

surajsharma commented:
@georgeciubotaru fixed all issues in last push, pls see

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/richLink.svelte:
```diff
@@ -0,0 +1,139 @@
+  class="flex border border-l4 bg-l2 rounded-xl w-full items-center h-24"
+  target={isHoldexLink ? '_self' : '_blank'}
+>
+  {#await promise}
+    <div class="h-full flex items-center">
+      <div
+        class="w-full rich-link-image-skeleton h-full rounded-xl bg-contain bg-no-repeat bg-center animation"
+      />
+      <div class="p-4 flex flex-col text-footnote justify-between gap-1 w-full overflow-hidden">
+        <div class="animation animate h-4 skeleton-desc" />
+        <div class="animation animate h-4" />
+        <div>
+          <a
+            title={' '}
+            href={item.data?.url}
+            target={isHoldexLink ? '_self' : '_blank'}
+            class={classes}
+            rel="noreferrer"
+          >
+            {truncated}
+            {#if !isHoldexLink}
+              <Icon icon={ArrowTopRightOnSquare} width={16} height={16} colorInherit />
+            {/if}
```
you cannot have `a` element inside an `a` element

georgeciubotaru reviewed (changes requested):
see my change

surajsharma commented:
> @surajsharma Your solution is not acceptable, please move to another PR.

@georgeciubotaru why is it not acceptable?

georgeciubotaru commented:
@surajsharma
1. Opened 3 weeks ago (more than 4hrs)
2.  My involvement time exceeded the actual total development time (4hrs)
3. The code is not production-ready yet

@markholdex, @zolotokrylin  FYI

zolotokrylin commented:
@surajsharma 

1. why does your PR take so long? 
2. how much more time do you need to complete PR? George can't baby sit here, he has other things to do.

surajsharma commented:
@zolotokrylin 

> 1. why does your PR take so long?

this was in draft for a long while i was working on other PRs

> 2. how much more time do you need to complete PR? George can't baby sit here, he has other things to do.

i understand he is busy, the last interaction here was fixing some of his comments on vercel, from my side it is good to go

georgeciubotaru commented:
@surajsharma So what is the ETA to finish this task?

I will review this PR after
- It will be pixel-perfect
- correct semantic

surajsharma commented:
@georgeciubotaru 

- here's an image of pixel-perfection

<img width="934" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/3ccf7894-2717-431e-93f6-5005452ec063">


- semantics are already fixed

zolotokrylin commented:
@georgeciubotaru please check the last time. If the work is done, then it is done. 


And let's just follow the principle we have discussed where we do once a week review on Friday and where we give 2 reviews max.

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/richLink.svelte:
```diff
@@ -0,0 +1,150 @@
+        <div>
+          <a
+            title={''}
+            href={item.data?.url}
```
you already have the `link` parent block that is clickable, you double the event handler.

the structure should be simple:

old structure example:
<img width="1038" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/8217518/06456ead-6b9b-4fcd-a678-b0f7d4995dcf">

surajsharma replied:
removed extra a tag @georgeciubotaru

georgeciubotaru commented on src/routes/(pages)/layout.scss:
```diff
@@ -19,3 +19,11 @@
   @apply min-w-[150px] max-w-[200px];
 }
+
+.rich-link-image {
```
move that style in the related component

surajsharma replied:
done @georgeciubotaru

georgeciubotaru commented:
Your implementation:
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/3f8a9683-3b4a-49b0-a2e5-c7be37e65a72)

Design:
![image](https://github.com/holdex/holdex-venture-studio/assets/8217518/ade00a39-b00a-4c1e-af55-dcc6e4a62ac8)

surajsharma commented:
> Your implementation: ![image](https://user-images.githubusercontent.com/8217518/267876992-3f8a9683-3b4a-49b0-a2e5-c7be37e65a72.png)
> 
> Design: ![image](https://user-images.githubusercontent.com/8217518/267877104-ade00a39-b00a-4c1e-af55-dcc6e4a62ac8.png)

this is pixel perfect overlay of figma export

![image](https://github.com/holdex/holdex-venture-studio/assets/5298624/b4ae55ed-455c-4f8d-be49-f960f6cc71b5)


@georgeciubotaru @zolotokrylin

surajsharma commented:
@georgeciubotaru 

<img width="671" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/63561045-90ce-440b-9bad-863e1f837d74">



design:
<img width="129" alt="image" src="https://github.com/holdex/holdex-venture-studio/assets/5298624/d07ff17c-f85e-47d2-8d4f-3e57898a3617">

georgeciubotaru commented:
@surajsharma 

1. The paddings are wrong (check for image padding)
2. The structure is not semantic
 - I've placed a screenshot with the old structure, check it and try to have the same one

surajsharma commented:
@georgeciubotaru 

> 2. The structure is not semantic

pls give a list of semantic elements you would like to see used here, there are many `divs` and `spans` so not exactly sure what you are asking for here, i will repace with semantic elements where i see fit

> 1. The paddings are wrong (check for image padding)

i have added screenshot with pixel perfect showing no gap in padding, here's another video showing dimensions are exactly as in design as of now



https://github.com/holdex/holdex-venture-studio/assets/5298624/b65e8b95-7c40-414b-8051-6f493ffcb883

surajsharma commented:
@georgeciubotaru i believe this is good to go as per your directions, i have added semantic elements, fixed padding

georgeciubotaru commented:
@surajsharma Please check the screenshot with the semantics from the old implementation.

In your case, you have an `a` element inside a `span` wrapper that has `clickHandler`,
-  case of click to the link you are doubling the event
- then entire block should be a link

georgeciubotaru commented:
@surajsharma Regarding image padding

You have a `figure` block that should have padding and the `rich-link-image` to fit the space inside it

surajsharma commented:
> @surajsharma Regarding image padding
> 
> You have a `figure` block that should have padding and the `rich-link-image` to fit the space inside it

i had to use some creative css here to ensure aspect ratio, for this reason, took out the `figure` and replaced it with a section, renamed css classes to remove confusion as to whats happenning

```
  style={`background-size:cover; background-position: center;
    background-repeat: no-repeat; background-image: url(${ogdata?.image[0]?.url});`}

``

surajsharma commented:
> @surajsharma Please check the screenshot with the semantics from the old implementation.
> 
> In your case, you have an `a` element inside a `span` wrapper that has `clickHandler`,
> -  case of click to the link you are doubling the event
> - then entire block should be a link

@georgeciubotaru let me check

surajsharma commented:
@georgeciubotaru all issues should be fixed now, apologies for the trouble with semantics

georgeciubotaru commented:
@surajsharma 

1. Please strictly follow the semantics from my screenshot (old implementation)
2. Title/href truncation should be done strictly in CSS (no JS functions)

surajsharma commented:
> @surajsharma
> 
> 1. Please strictly follow the semantics from my screenshot (old implementation)
> 2. Title/href truncation should be done strictly in CSS (no JS functions)

@georgeciubotaru please leave comments in code only, otherwise i don't understand so well

surajsharma commented on src/lib/components/BodyRenderer/Blocks/richLink.svelte:
```diff
@@ -0,0 +1,144 @@
+    'cursor-pointer inline-block text-footnote block underline underline-offset-4 bg-accent1-default/15 text-accent1-default transition-colors hover:bg-accent1-default/25 focus:bg-accent1-default/25';
+
+  let truncateUrl = (url: string) => {
+    let textWithoutPrefix = url.replace(/^https?:\/\//, '');
```
@georgeciubotaru 

> 2. Title/href truncation should be done strictly in CSS (no JS functions)

is it possible to truncate with regex match in css? because elsewhere in the code we seem to be making use of regex for href truncation

georgeciubotaru reviewed (changes requested):
see my comment

georgeciubotaru commented on src/lib/components/BodyRenderer/Blocks/richLink.svelte:
```diff
@@ -0,0 +1,144 @@
+  };
+</script>
+
+<span
```
@surajsharma If you check the old structure, you see that the 

`span` wrapper doesn't have any click handler, and the first child is an `a` tag with a link
